src/lower/default.rs

```diff
@@ -0,0 +1,80 @@
+use ir::*;
+use solve::infer::InferenceTable;
+use cast::Cast;
+
+impl Program {
+    pub(super) fn add_default_impls(&mut self) {
+        // For each auto trait `MyAutoTrait` and for each struct/type `MyStruct`
+        for auto_trait in self.trait_data.values().filter(|t| t.binders.value.auto) {
+            for struct_datum in self.struct_data.values() {
+
+                // `MyStruct: MyAutoTrait`
+                let trait_ref = TraitRef {
+                    trait_id: auto_trait.binders.value.trait_ref.trait_id,
+                    parameters: vec![
+                        ParameterKind::Ty(Ty::Apply(struct_datum.binders.value.self_ty.clone()))
+                    ]
+                };
+
+                // If a positive or negative impl is already provided for a type family
+                // which includes `MyStruct`, we do not generate a default impl.
+                if self.provide_impl_for(trait_ref.clone(), struct_datum) {
+                    continue;
+                }
+
+                self.default_impl_data.push(DefaultImplDatum {
+                    binders: Binders {
+                        binders: struct_datum.binders.binders.clone(),
+                        value: DefaultImplDatumBound {
+                            trait_ref,
+                            accessible_tys: struct_datum.binders.value.fields.clone(),
+                        }
+                    }
+                });
+            }
+        }
+    }
+
+    fn provide_impl_for(&self, trait_ref: TraitRef, struct_datum: &StructDatum) -> bool {
```

Comment:
Right, actually I was reading this like "self provide_impl_for struct", but I agree this can be confusing.